### PR TITLE
Feat/dbaas

### PR DESCRIPTION
**Pull request by IuliaBlindu**
## What does this fix or implement?

- created bundle for ionoscloud and dbaas clients 
- implemented resource for dbaas cluster with all CRUD operations 
- created tests for both create and update with updates on all updateable fields

## Checklist

- [x] Tests added or updated
- [x] Documentation updated
- [ ] Changelog updated and version incremented
- [ ] Github Issue linked if any

### CODE REVIEW

File: ionoscloud/data_source_dbaas_pgsql_cluster.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+					},
+				},
+			},
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func dataSourceDbaasPgSqlReadCluster(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(SdkBundle).DbaasClient
```
Comment (mflorin) [inline ionoscloud/data_source_dbaas_pgsql_cluster.go:128]: Lets externalize the read logic to be able to reuse it in the import section and in data sources.

File: ionoscloud/data_source_dbaas_pgsql_versions.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+				versions = append(versions, *version.Name)
+			}
+		}
+		err := d.Set("postgres_versions", versions)
+		if err != nil {
+			diags := diag.FromErr(fmt.Errorf("error while setting postgres_versions: %s", err))
+			return diags
+		}
+	}
+
```
Comment (mflorin) [inline ionoscloud/data_source_dbaas_pgsql_versions.go:69]: Same thing, lets extract the code we can reuse in import and data sources.

File: go.mod [hunk 3]
```diff
@@ -24,3 +25,5 @@ require (
 	google.golang.org/appengine v1.6.7 // indirect
 	gopkg.in/yaml.v3 v3.0.0-20200615113413-eeeca48fe776
 )
+
+replace github.com/ionos-cloud/sdk-go-autoscaling => ../../../../../sdk/dbaas-sdk
```
Comment (cristiGuranIonos) [inline go.mod:30]: will this work?

File: ionoscloud/data_source_dbaas_pgsql_cluster.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	dbaasService "github.com/ionos-cloud/terraform-provider-ionoscloud/services/dbaas"
+)
+
+func dataSourceDbaasPgSqlCluster() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceDbaasPgSqlReadCluster,
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:         schema.TypeString,
+				Description:  "The is of your cluster.",
```
Comment (cristiGuranIonos) [inline ionoscloud/data_source_dbaas_pgsql_cluster.go:20]: id

File: ionoscloud/data_source_dbaas_pgsql_cluster.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+							Computed:    true,
+						},
+						"lan_id": {
+							Type:        schema.TypeString,
+							Description: "The LAN to connect your cluster to.",
+							Computed:    true,
+						},
+						"cidr": {
+							Type:        schema.TypeString,
+							Description: "The IP and subnet for the database.\n          Note the following unavailable IP ranges:\n          10.233.64.0/18\n          10.233.0.0/18\n          10.233.114.0/24",
```
Comment (cristiGuranIonos) [inline ionoscloud/data_source_dbaas_pgsql_cluster.go:77]: no /n i think

File: ionoscloud/data_source_dbaas_pgsql_cluster.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		if clusters.Items != nil && len(*clusters.Items) > 0 {
+			cluster = (*clusters.Items)[0]
+		} else {
+			return diag.FromErr(errors.New("dbaas cluster not found"))
+		}
+	}
+
+	if cluster.Id != nil {
+		if err := d.Set("id", *cluster.Id); err != nil {
+			return nil
```
Comment (cristiGuranIonos) [inline ionoscloud/data_source_dbaas_pgsql_cluster.go:195]: return nil, or err?

File: ionoscloud/resource_dbaas_pgsql_cluster.go [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+}
+
+func resourceDbaasPgSqlClusterRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
+	client := meta.(SdkBundle).DbaasClient
+
+	cluster, apiResponse, err := client.GetCluster(ctx, d.Id())
+
+	if err != nil {
+		if apiResponse != nil && apiResponse.StatusCode == 404 {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_dbaas_pgsql_cluster.go:225]: should we check apiResponse.Response?

File: ionoscloud/resource_lan.go [hunk 8]
```diff
@@ -327,19 +330,32 @@ func lanAvailable(ctx context.Context, client *ionoscloud.APIClient, d *schema.R
...
 	if err != nil {
 		if apiResponse != nil && apiResponse.Response != nil && apiResponse.StatusCode == 404 {
 			return true, nil
 		}
-		return true, fmt.Errorf("error checking LAN deletion status: %s", err)
+		return false, fmt.Errorf("error checking LAN deletion status: %s", err)
+	}
+
+	log.Printf("[INFO] LAN %s not deleted yet deleted from the datacenter %s", d.Id(), dcId)
+	log.Printf("[INFO] Current deletion status for LAN %s: %+v", d.Id(), *rsp.Metadata.State)
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_lan.go:346]: maybe check rsp metadata

File: ionoscloud/resource_dbaas_pgsql_cluster.go [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+}
+
+func resourceDbaasPgSqlClusterRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
+	client := meta.(SdkBundle).DbaasClient
+
+	cluster, apiResponse, err := client.GetCluster(ctx, d.Id())
+
+	if err != nil {
+		if apiResponse != nil && apiResponse.Response != nil && apiResponse.StatusCode == 404 {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_dbaas_pgsql_cluster.go:225]: maybe easier to use the httpNotFound function, to make sure we don't miss any api.Response nil checks

File: ionoscloud/resource_dbaas_pgsql_cluster.go [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+
+func resourceDbaasPgSqlClusterImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	client := meta.(SdkBundle).DbaasClient
+
+	clusterId := d.Id()
+
+	dbaasCluster, apiResponse, err := client.GetCluster(ctx, clusterId)
+
+	if err != nil {
+		if apiResponse != nil && apiResponse.Response != nil && apiResponse.StatusCode == 404 {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_dbaas_pgsql_cluster.go:340]: aren't the error messages reversed here?

File: ionoscloud/resource_dbaas_pgsql_cluster.go [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+	return []*schema.ResourceData{d}, nil
+}
+
+func dbaasClusterReady(ctx context.Context, client *dbaasService.Client, d *schema.ResourceData) (bool, error) {
+	subjectCluster, _, err := client.GetCluster(ctx, d.Id())
+
+	if err != nil {
+		return true, fmt.Errorf("error checking dbaas cluster status: %s", err)
+	}
+	// Removed this part since there are still problems with the clusters being unstable (failing for a short time and then recovering)
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_dbaas_pgsql_cluster.go:362]: maybe add a todo here in the code so it stands out

**Pull request merged**